File: platform/wasm/web/src/render.js
'use strict'

import {runtime} from "./runtime";

/**
 * Convert a \#AARRGGBB color to CSS color.
 *
 * @param {Number} color the \#AARRGGBB color
 */
function toCSSColor(color) {
  const css = ((color << 8) >>> 0 | color >>> 24) >>> 0;
  let str = css.toString(16);
  if (str.length < 8) {
    for (let i = 8 - str.length; i > 0; i--) {
      str = '0' + str;
    }
  }
  return `#${str}`;
}

/**
 * Array to get command arguments.
 *
 * @type {Float32Array}
 */
const argBuf = new Float32Array(8);

/**
 * Create a Render to paint formulas. You MUST call {@link Render.release} to free
 * memory after this render is unused.
 *
 * @param {Number} nativeRender the pointer of native render parsed by the engine
 * @param {boolean} isLittleEndian if the byte order is little endian
 * @constructor
 */
function Render(nativeRender, isLittleEndian) {
  const _nativeRender = nativeRender;
  const _isLittleEndian = isLittleEndian;

  /**
   * Get the paintable formula width.
   *
   * @return {Number}
   */
  this.getWidth = function () {
    return runtime._microtex_getRenderWidth(_nativeRender);
  }

  /**
   * Get the height of the render.
   *
   * @return {Number}
   */
  this.getHeight = function () {
    return runtime._microtex_getRenderHeight(_nativeRender);
  }

  /**
   * Get the depth below the baseline, in positive.
   *
   * @return {Number}
   */
  this.getDepth = function () {
    return runtime._microtex_getRenderDepth(_nativeRender);
  }

  const capMap = {0: "butt", 1: "round", 2: "square"};
  const joinMap = {0: "bevel", 1: "round", 2: "miter"};

  /**
   * Draw the formula on point (x, y).
   *
   * @param {CanvasRenderingContext2D} ctx the graphical (2D) context
   * @param {Number} x the x coordinate
   * @param {Number} y the y coordinate
   */
  this.draw = function (ctx, x, y) {
    const ptr = runtime._microtex_getDrawingData(_nativeRender, x, y);
    // make a view to iterate drawing commands
    const v = new DataView(runtime.HEAPU8.buffer);
    let offset = 0;

    function getU8() {
      const x = v.getUint8(ptr + offset);
      offset += 1;
      return x;
    }

    function getU32() {
      const x = v.getUint32(ptr + offset, _isLittleEndian);
      offset += 4;
      return x;
    }

    function getF32() {
      const x = v.getFloat32(ptr + offset, _isLittleEndian);
      offset += 4;
      return x;
    }

    function getF32s(cnt) {
      for (let j = 0; j < cnt; j++) {
        argBuf[j] = v.getFloat32(ptr + offset, _isLittleEndian);
        offset += 4;
      }
      return argBuf;
    }

    /** Make a round rectangle. */
    function roundRect(x, y, w, h, rx, ry) {
      const r = Math.max(rx, ry);
      const d = Math.PI / 180;
      ctx.beginPath();
      ctx.arc(x + r, y + r, r, 180 * d, 270 * d);
      ctx.arc(x + w - r, y + r, r, -90 * d, 0);
      ctx.arc(x + w - r, y + h - r, r, 0, 90 * d);
      ctx.arc(x + r, y + h - r, r, 90 * d, 180 * d);
      ctx.closePath();
    }

    const len = getU32();

    let sx = 1;
    let sy = 1;
    while (offset < len) {
      const cmd = getU8();
      switch (cmd) {
        case 0: // setColor
          const color = getU32();
          const css = toCSSColor(color);
          ctx.fillStyle = css;
          ctx.strokeStyle = css;
          break;
        case 1: // setStroke
          const lineWidth = getF32();
          const miterLimit = getF32();
          const cap = getU32();
          const join = getU32();
          ctx.lineWidth = lineWidth;
          ctx.miterLimit = miterLimit;
          ctx.lineCap = capMap[cap];
          ctx.lineJoin = joinMap[join];
          break;
        case 2: // translate
          const t = getF32s(2);
          ctx.translate(t[0], t[1]);
          break;
        case 3: // scale
          const s = getF32s(2);
          sx *= s[0];
          sy *= s[1];
          ctx.scale(s[0], s[1]);
          break;
        case 4: // rotate
          const r = getF32s(3);
          ctx.translate(r[1], r[2]);
          ctx.rotate(r[0]);
          ctx.translate(-r[1], -r[2]);
          break;
        case 5: // reset
          ctx.setTransform(1, 0, 0, 1, 0, 0);
          break;
        case 6: // moveTo
          const m = getF32s(2);
          ctx.moveTo(m[0], m[1]);
          break;
        case 7: // lineTo
          const l = getF32s(2);
          ctx.lineTo(l[0], l[1]);
          break;
        case 8: // cubicTo
          const c = getF32s(6);
          ctx.bezierCurveTo(c[0], c[1], c[2], c[3], c[4], c[5]);
          break;
        case 9: // quadTo
          const q = getF32s(4);
          ctx.quadraticCurveTo(q[0], q[1], q[2], q[3]);
          break;
        case 10: // closePath
          ctx.closePath();
          break;
        case 11: // fillPath
          ctx.fill();
          ctx.beginPath();
          break;
        case 12: // drawLine
          const dl = getF32s(4);
          ctx.beginPath();
          ctx.moveTo(dl[0], dl[1]);
          ctx.lineTo(dl[2], dl[3]);
          ctx.stroke();
          break;
        case 13: // drawRect
          const dr = getF32s(4);
          ctx.strokeRect(dr[0], dr[1], dr[2], dr[3]);
          break;
        case 14: // fillRect
          const fr = getF32s(4);
          ctx.fillRect(fr[0], fr[1], fr[2], fr[3]);
          break;
        case 15: // drawRoundRect
          const rr = getF32s(6);
          roundRect(rr[0], rr[1], rr[2], rr[3], rr[4], rr[5]);
          ctx.stroke();
          break;
        case 16: // fillRoundRect
          const rf = getF32s(6);
          roundRect(rf[0], rf[1], rf[2], rf[3], rf[4], rf[5]);
          ctx.fill();
          break;
        case 17: // beginPath
          ctx.beginPath();
          break;
        case 18: // drawTextLayout
          const id = getU32();
          const xy = getF32s(2);
          const old = ctx.font;
          const value = runtime.textLayout.map.get(id);
          ctx.font = value.font;
          ctx.fillText(value.text, xy[0], xy[1]);
          ctx.font = old;
          break;
        case 19: // setDash
          const hasDash = getU8() === 1;
          const dash = hasDash ? [5 / sx, 5 / sx] : [];
          ctx.setLineDash(dash);
          break;
        default:
          // invalid drawing command
          break;
      }
    }
    runtime._free(ptr);
  }

  /** Release the native render. */
  this.release = function () {
    runtime._microtex_deleteRender(nativeRender);
  }
}

export {Render}

File: platform/flutter/microtex/lib/src/render.dart
import 'dart:convert';
import 'dart:ffi';
import 'dart:typed_data';
import 'dart:ui';

import 'package:flutter/rendering.dart';
import 'package:microtex/microtex.dart';
import 'package:microtex/src/bindings.dart';
import 'package:microtex/src/pathcache.dart';
import 'package:microtex/src/textlayout.dart';

class Render {
  final NativeBindings _bindings;
  final Pointer _nativeInstance;
  final Endian _endian;

  /// If filter is enabled, using a bluring to simulate the 'grid fitting' on glyhs
  /// drawing to get a much clear rendering, but it is expensive, especially for
  /// complex formulas.
  /// See [https://github.com/flutter/flutter/issues/104017].
  Blur? blur;

  Render(this._bindings, this._nativeInstance, bool isLittleEndian)
      : _endian = isLittleEndian ? Endian.little : Endian.big;

  /// The width of this render.
  int get width => _bindings.getRenderWidth(_nativeInstance);

  /// The height of this render, equals to ascent + descent (or namely [depth],
  /// distance below the baseline).
  int get height => _bindings.getRenderHeight(_nativeInstance);

  /// The distance below the baseline, in positive.
  int get depth => _bindings.getRenderDepth(_nativeInstance);

  /// Test if this render is split to multi-lines.
  bool get isSplit => _bindings.isRenderSplit(_nativeInstance);

  void release() {
    _bindings.deleteRender(_nativeInstance);
  }

  /// Set the text size to draw. The dimension ([width], [height], [depth]) will be
  /// changed following the text size change.
  void setTextSize(double textSize) {
    _bindings.setRenderTextSize(_nativeInstance, textSize);
  }

  /// Set the foreground color to draw.
  void setColor(int color) {
    _bindings.setRenderForground(_nativeInstance, color);
  }

  /// If this render is use path to render glyphs.
  ///
  /// Unlike the [MicroTeX.isRenderGlyphUsePath], this is for every particular renders.
  bool isRenderGlyphUsePath = false;

  static const _caps = [StrokeCap.butt, StrokeCap.round, StrokeCap.square];
  static const _joins = [StrokeJoin.bevel, StrokeJoin.round, StrokeJoin.miter];

  final _paint = Paint()..isAntiAlias = true;
  final _argBuf = Float32List(8);

  void draw(Canvas canvas) {
    print("Render.draw");
    // it is OK to change the value of the property [MicroTeX.isRenderGlyphUsePath] when
    // drawing every renders, since draw runs in UI thread.
    final prevIsRenderGlyphUsePath = MicroTeX().isRenderGlyphUsePath();
    MicroTeX().setRenderGlyphUsePath(isRenderGlyphUsePath);
    // get the drawing commands from native render
    final drawingData = _bindings.getDrawingData(_nativeInstance, 0, 0);
    // the first 4 bytes is the total count of byte used
    final n = drawingData.asTypedList(4);
    final len = n.buffer.asByteData().getUint32(0, _endian);

    // make a view to iterate over the drawing commands
    final v = drawingData.asTypedList(len).buffer.asByteData();
    var offset = 0;

    int getU8() {
      final x = v.getUint8(offset);
      offset += 1;
      return x;
    }

    T getNum<T>(int size, T Function(int offset) f) {
      final x = f(offset);
      offset += size;
      return x;
    }

    String getString() {
      final utf8 = <int>[];
      var byte = getU8();
      while (byte != 0) {
        utf8.add(byte);
        byte = getU8();
      }
      return const Utf8Decoder().convert(utf8);
    }

    int getU16() => getNum(2, (offset) => v.getUint16(offset, _endian));
    int getI32() => getNum(4, (offset) => v.getInt32(offset, _endian));
    int getU32() => getNum(4, (offset) => v.getUint32(offset, _endian));
    double getF32() => getNum(4, (offset) => v.getFloat32(offset, _endian));

    Float32List getF32s(int cnt) {
      for (var i = 0; i < cnt; i++) {
        _argBuf[i] = v.getFloat32(offset, _endian);
        offset += 4;
      }
      return _argBuf;
    }

    // consume the first 4 bytes
    getU32();
    var sx = 1.0, sy = 1.0;
    var path = Path();

    var fontFamily = "";
    var fontSize = 0.0;

    while (offset < len) {
      final cmd = getU8();
      switch (cmd) {
        case 0: // setColor
          final color = getU32();
          _paint.color = Color(color);
          break;
        case 1: // setStroke
          final lineWidth = getF32();
          final miterLimit = getF32();
          final cap = getU32();
          final join = getU32();
          _paint
            ..strokeWidth = lineWidth
            ..strokeMiterLimit = miterLimit
            ..strokeCap = _caps[cap]
            ..strokeJoin = _joins[join];
          break;
        case 2: // setDash
          final hasDash = getU8() == 1;
          final dash = hasDash ? [5 / sx, 5 / sx] : [];
          // FIXME
          // Flutter does not support dashing lines
          break;
        case 3: // setFont
          // TODO
          final family = getString();
          fontFamily = family;
          break;
        case 4: // setFontSize
          // TODO
          final size = getF32();
          fontSize = size;
          break;
        case 5: // translate
          final t = getF32s(2);
          canvas.translate(t[0], t[1]);
          break;
        case 6: // scale
          final s = getF32s(2);
          sx *= s[0];
          sy *= s[1];
          canvas.scale(s[0], s[1]);
          break;
        case 7: // rotate
          final r = getF32s(3);
          canvas.translate(r[1], r[2]);
          canvas.rotate(r[0]);
          canvas.translate(-r[1], -r[2]);
          break;
        case 8: // reset
          canvas.transform(Matrix4.identity().storage);
          break;
        case 9: // drawGlyph
          // TODO
          final glyph = getU16();
          final xy = getF32s(2);
          final run = GlyphRun(
            glyphs: Uint16List.fromList([glyph]),
            positions: [Offset(xy[0], xy[1])],
            fontFamily: fontFamily,
            fontSize: fontSize,
          );
          canvas.drawGlyphRun(run, const Offset(0, 0), _paint);
          break;
        case 10: // beginPath
          final pid = getI32();
          final p = PathCache()[pid];
          path = p ?? Path();
          break;
        case 11: // moveTo
          final m = getF32s(2);
          path.moveTo(m[0], m[1]);
          break;
        case 12: // lineTo
          final l = getF32s(2);
          path.lineTo(l[0], l[1]);
          break;
        case 13: // cubicTo
          final c = getF32s(6);
          path.cubicTo(c[0], c[1], c[2], c[3], c[4], c[5]);
          break;
        case 14: // quadTo
          final q = getF32s(4);
          path.quadraticBezierTo(q[0], q[1], q[2], q[3]);
          break;
        case 15: // closePath
          path.close();
          break;
        case 16: // fillPath
          final pid = getI32();
          _paint.style = PaintingStyle.fill;
          final b = blur;
          final p = b == null ? _paint : (_paint.copy()..maskFilter = MaskFilter.blur(b.style, b.sigma / sx));
          PathCache()[pid] = path;
          canvas.drawPath(path, p);
          break;
        case 17: // drawLine
          final dl = getF32s(4);
          canvas.drawLine(Offset(dl[0], dl[1]), Offset(dl[2], dl[3]), _paint);
          break;
        case 18: // drawRect
          final dr = getF32s(4);
          _paint.style = PaintingStyle.stroke;
          canvas.drawRect(Rect.fromLTWH(dr[0], dr[1], dr[2], dr[3]), _paint);
          break;
        case 19: // fillRect
          final fr = getF32s(4);
          _paint.style = PaintingStyle.fill;
          canvas.drawRect(Rect.fromLTWH(fr[0], fr[1], fr[2], fr[3]), _paint);
          break;
        case 20: // drawRoundRect
          final rr = getF32s(6);
          _paint.style = PaintingStyle.stroke;
          canvas.drawRRect(
            RRect.fromRectXY(Rect.fromLTWH(rr[0], rr[1], rr[2], rr[3]), rr[4], rr[5]),
            _paint,
          );
          break;
        case 21: // fillRoundRect
          final rf = getF32s(6);
          _paint.style = PaintingStyle.fill;
          canvas.drawRRect(
            RRect.fromRectXY(Rect.fromLTWH(rf[0], rf[1], rf[2], rf[3]), rf[4], rf[5]),
            _paint,
          );
          break;
        case 22: // drawTextLayout
          final id = getU32();
          final xy = getF32s(2);
          TextLayout.draw(id, canvas, xy[0], xy[1], _paint.color);
          break;
        default:
          // invalid drawing command
          break;
      }
    }
    // delete the drawing data
    _bindings.deleteDrawingData(drawingData);
    // reset the property
    MicroTeX().setRenderGlyphUsePath(prevIsRenderGlyphUsePath);
  }

  @override
  String toString() {
    return "Render{$width, $height, $depth}";
  }
}

extension _PaintExt on Paint {
  Paint copy() {
    return Paint()
      ..isAntiAlias = isAntiAlias
      ..color = color
      ..blendMode = blendMode
      ..style = style
      ..strokeWidth = strokeWidth
      ..strokeCap = strokeCap
      ..strokeJoin = strokeJoin
      ..strokeMiterLimit = strokeMiterLimit
      ..maskFilter = maskFilter
      ..filterQuality = filterQuality
      ..shader = shader
      ..colorFilter = colorFilter
      ..imageFilter = imageFilter
      ..invertColors = invertColors;
  }
}
